From d96ae86ed1e1219716c056fd4414a980924c394e Mon Sep 17 00:00:00 2001
From: Yumi Yukimura <me.cafebabe@gmail.com>
Date: Tue, 8 Sep 2026 17:05:59 +0800
Subject: [PATCH 1/2] [nl80211] Fix random MAC scan feature parsing

NL80211 feature constants represent bit indexes, but were used directly
as masks. This falsely enabled random MAC scanning when unrelated feature
bits were set.

Convert the feature indexes to masks before checking them.

Assisted-by: OpenCode/gpt-5.6-sol
Change-Id: I70b358dd8c1dd295b26d7064952f7ec2e1c72237
---
 .../server/wifi/nl80211/Nl80211Utils.java     |  4 +--
 .../server/wifi/nl80211/Nl80211UtilsTest.java | 30 +++++++++++++++++++
 2 files changed, 32 insertions(+), 2 deletions(-)

diff --git a/service/java/com/android/server/wifi/nl80211/Nl80211Utils.java b/service/java/com/android/server/wifi/nl80211/Nl80211Utils.java
index c90f275dcd..f37764b19b 100644
--- a/service/java/com/android/server/wifi/nl80211/Nl80211Utils.java
+++ b/service/java/com/android/server/wifi/nl80211/Nl80211Utils.java
@@ -2266,9 +2266,9 @@ public class Nl80211Utils {
 
     private WiphyFeatures createWiphyFeatures(int featureFlags, byte[] extFeatureFlagsBytes) {
         boolean supportsRandomMacOneShotScan =
-                (featureFlags & NL80211_FEATURE_SCAN_RANDOM_MAC_ADDR) != 0;
+                (featureFlags & (1 << NL80211_FEATURE_SCAN_RANDOM_MAC_ADDR)) != 0;
         boolean supportsRandomMacSchedScan =
-                (featureFlags & NL80211_FEATURE_SCHED_SCAN_RANDOM_MAC_ADDR) != 0;
+                (featureFlags & (1 << NL80211_FEATURE_SCHED_SCAN_RANDOM_MAC_ADDR)) != 0;
 
         boolean supportsLowSpanOneShotScan = isExtFeatureFlagSet(extFeatureFlagsBytes,
                 NL80211_EXT_FEATURE_LOW_SPAN_SCAN);
diff --git a/service/tests/wifitests/src/com/android/server/wifi/nl80211/Nl80211UtilsTest.java b/service/tests/wifitests/src/com/android/server/wifi/nl80211/Nl80211UtilsTest.java
index 5454abc5ac..098a875264 100644
--- a/service/tests/wifitests/src/com/android/server/wifi/nl80211/Nl80211UtilsTest.java
+++ b/service/tests/wifitests/src/com/android/server/wifi/nl80211/Nl80211UtilsTest.java
@@ -580,6 +580,36 @@ public class Nl80211UtilsTest {
         assertEquals(16, info.scanCapabilities.maxNumScanSsids);
     }
 
+    @Test
+    public void testParseWiphyInfo_unrelatedFeatureFlags_doNotEnableRandomMacScan() {
+        GenericNetlinkMsg msg = createBasicWiphyInfoMsg();
+        msg.addAttribute(createWiphyBandsAttribute());
+        msg.addAttribute(new StructNlAttr(NL80211_ATTR_FEATURE_FLAGS,
+                (1 << NetlinkConstants.NL80211_FEATURE_SK_TX_STATUS)
+                        | (1 << NetlinkConstants.NL80211_FEATURE_HT_IBSS)));
+
+        Nl80211Utils.WiphyInfo info = mNl80211Utils.parseWiphyInfo(List.of(msg));
+
+        assertNotNull(info);
+        assertFalse(info.wiphyFeatures.supportsRandomMacOneShotScan);
+        assertFalse(info.wiphyFeatures.supportsRandomMacSchedScan);
+    }
+
+    @Test
+    public void testParseWiphyInfo_randomMacFeatureFlags_enableRandomMacScan() {
+        GenericNetlinkMsg msg = createBasicWiphyInfoMsg();
+        msg.addAttribute(createWiphyBandsAttribute());
+        msg.addAttribute(new StructNlAttr(NL80211_ATTR_FEATURE_FLAGS,
+                (1 << NetlinkConstants.NL80211_FEATURE_SCAN_RANDOM_MAC_ADDR)
+                        | (1 << NetlinkConstants.NL80211_FEATURE_SCHED_SCAN_RANDOM_MAC_ADDR)));
+
+        Nl80211Utils.WiphyInfo info = mNl80211Utils.parseWiphyInfo(List.of(msg));
+
+        assertNotNull(info);
+        assertTrue(info.wiphyFeatures.supportsRandomMacOneShotScan);
+        assertTrue(info.wiphyFeatures.supportsRandomMacSchedScan);
+    }
+
     @Test
     public void testParseWiphyInfo_antennas_success() {
         GenericNetlinkMsg msg = createBasicWiphyInfoMsg();
-- 
2.47.3

